src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java

```diff
@@ -3565,18 +3565,74 @@ private String lookupFQNs(String javacType) {
+                if (classAndPackageMap.containsKey(JavaParserUtil.erase(type.asString()))) {
+                  fullyQualifiedName.append(classAndPackageMap.get(type.asString()));
+                  fullyQualifiedName.append(".");
+                } else if (!type.getTypeArguments().isPresent()) {
+                  // This type is in the same package and doesn't contain any type parameters
                   return super.visit(type, p);
                 }
+
+                NodeList<Type> typeArguments = type.getTypeArguments().orElse(null);
+
+                if (typeArguments != null) {
+                  fullyQualifiedName.append(
+                      type.asString().substring(0, type.asString().indexOf('<') + 1));
+
+                  for (int i = 0; i < typeArguments.size(); i++) {
+                    Type typeArgument = typeArguments.get(i);
+                    lookupTypeArgumentFQN(fullyQualifiedName, typeArgument);
+
+                    if (i < typeArguments.size() - 1) {
+                      fullyQualifiedName.append(", ");
+                    }
+                  }
+                  fullyQualifiedName.append(">");
+                } else {
+                  fullyQualifiedName.append(type.asString());
+                }
+
+                return StaticJavaParser.parseClassOrInterfaceType(fullyQualifiedName.toString());
               }
             },
             null);
     return typeVarDecl + parsedJavac.toString();
   }
 
+  /**
+   * Helper method for lookupFQNs which adds the fully qualified type argument to
+   * fullyQualifiedName.
+   *
+   * @param fullyQualifiedName the fully qualified name to build
+   * @param typeArgument the type argument to lookup
+   */
+  private void lookupTypeArgumentFQN(StringBuilder fullyQualifiedName, Type typeArgument) {
+    String erased = JavaParserUtil.erase(typeArgument.asString());
+    if (classAndPackageMap.containsKey(erased)) {
+      fullyQualifiedName
+          .append(classAndPackageMap.get(erased))
+          .append(".")
+          .append(typeArgument.asString());
+    } else if (JavaLangUtils.isJavaLangName(erased)) {
+      // Keep java.lang type arguments as is (Integer, String, etc.)
+      fullyQualifiedName.append(typeArgument.asString());
+    } else if (typeArgument.isWildcardType()) {
+      WildcardType asWildcardType = typeArgument.asWildcardType();
+
+      if (asWildcardType.getSuperType().isPresent()) {
+        fullyQualifiedName.append("? super ");
+        lookupTypeArgumentFQN(fullyQualifiedName, asWildcardType.getSuperType().get());
+      } else if (asWildcardType.getExtendedType().isPresent()) {
+        fullyQualifiedName.append("? extends ");
+        lookupTypeArgumentFQN(fullyQualifiedName, asWildcardType.getExtendedType().get());
+      }
+    } else {
+      // If it's not imported, it's probably in the same package
+      fullyQualifiedName.append(currentPackage).append(".").append(typeArgument.toString());
```

Comment:
What if the type argument was already fully-qualified in the input. For example, what if we added the following method to your test case:
```
    public Bar<com.foo.InOtherPackage2> foo() {
        return Bar.getOtherPackage2();
    }
```

I think the current implementation will hit this line and append the current package to the front of the name, resulting in something like `"com.example.com.foo.InOtherPackage2"`. This would probably be easy enough to fix by just checking if the name is fully-qualified already before you call this method.

Regardless, please add the test case.

Reply:
@kelloggm I looked into this, and unfortunately it's more complicated than that. The `lookupFQN()` method, in this case, is called after `JavaTypeCorrect`, which gets rid of the fully qualified name. Since it's not imported, I can't resolve the fully qualified name through `classAndPackageMap`. I also don't think we can simply add the types to `classAndPackageMap` in `visit(ClassOrInterfaceType)`, since there could be multiple classes with the same simple name (i.e. `com.example.Foo` and `com.other.Foo`). Do you have any ideas on how to solve this?

Reply:
In the general case I think you're right that there's no unambiguous way to handle this.

I don't think handling this case correctly is critical at the moment. Let's move the `InOtherPackage2` version of the test into a separate test runner, allow that test runner to fail intentionally, and then add a TODO in `lookupFQN` with your notes about this problem. Then we can merge this PR, which does improve the current situation even if it is imperfect.

Then, we can come back to this problem later when we're not worried about a deadline at the end of the week :)